aws_sagemaker_studio/frameworks/pytorch_horovod_mpi/pytorch_horovod_mpi.ipynb

```diff
@@ -0,0 +1,407 @@
+    "grid = torchvision.utils.make_grid(images)\n",
+    "plt.imshow(grid.permute(1, 2, 0))  \n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Uploading the Data\n",
+    "\n",
+    "We use the `sagemaker.s3.S3Uploader.upload` function to upload our datasets to an S3 location. The return value, `inputs`, identifies the location of our data.\n",
+    "\n",
+    "To upload our data to S3, we need to specify an S3 bucket and prefix. These should be within the same region as the Notebook, training, and hosting."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sagemaker\n",
+    "from sagemaker.s3 import S3Uploader\n",
+    "\n",
+    "sagemaker_session = sagemaker.Session()\n",
+    "bucket = sagemaker_session.default_bucket()\n",
+    "prefix = 'pytorch-horovod-mpi-example'\n",
+    "inputs = S3Uploader.upload('data', 's3://{}/{}/data'.format(bucket, prefix))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Training"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Entrypoint Script\n",
+    "\n",
+    "In order to use Horovod with OpenMPI, we use a shell script as the training entry point. The shell script invokes our Python training script as part of an Open MPI command. You can learn more about running Horovod with MPI at https://horovod.readthedocs.io/en/stable/mpirun.html.\n",
+    "\n",
+    "Here's what the shell script looks like:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!cat src/train.sh"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We're using 4 GPUs because that's the number of GPUs on a ml.p2.8xlarge instance, but if you're using a different instance type you may need to change the value of the `-np` parameter."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Training Script"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Horovod-specific Code\n",
+    "\n",
+    "As you read the training script below, take note of these Horovod-specific modifications:\n",
+    "\n",
+    "* We start off by running `hvd.init()` and pinning each GPU to a single process\n",
+    "* While constructing the `torch.optim.SGD` optimizer, we scale the learning rate by the number of workers (if you're wondering why we do this, see [this paper](https://arxiv.org/abs/1706.02677))\n",
+    "* We wrap `torch.optim.SGD` with `hvd.DistributedOptimized` and broadcast the initial model state to the other processes\n",
+    "* We gaurd the `save` function with `hvd.local_rank() != 0` to ensure the model is saved only on worker 0. This is done to prevent other workers from corrupting the checkpoint\n",
```

Comment:
```suggestion
    "* We guard the `save` function with `hvd.local_rank() != 0` to ensure the model is saved only on worker 0. This is done to prevent other workers from corrupting the checkpoint\n",
```